src/main/java/net/fabricmc/loom/providers/MappingsProvider.java

```diff
@@ -66,50 +112,133 @@ public void provide(DependencyInfo dependency, Project project, LoomGradleExtens
+		this.mappingsName = StringUtils.removeSuffix(dependency.getDependency().getGroup() + "." + dependency.getDependency().getName(), "-unmerged");
+		if (this.isV2) mappingsName += "-v2";
 
 		Version mappingsVersion = new Version(version);
 		this.minecraftVersion = mappingsVersion.getMinecraftVersion();
 		this.mappingsVersion = mappingsVersion.getMappingsVersion();
 
 		initFiles(project);
 
-		if (!MAPPINGS_DIR.exists()) {
-			MAPPINGS_DIR.mkdir();
+		Files.createDirectories(mappingsDir);
+		Files.createDirectories(mappingsStepsDir);
+
+		if (!tinyMappings.exists()) {
+			storeMappings(project, minecraftProvider, mappingsJar);
 		}
 
-		if (!MAPPINGS_TINY_BASE.exists() || !MAPPINGS_TINY.exists()) {
-			if (!MAPPINGS_TINY_BASE.exists()) {
-				project.getLogger().lifecycle(":extracting " + mappingsJar.getName());
+		mappedProvider = new MinecraftMappedProvider();
+		mappedProvider.initFiles(project, minecraftProvider, this);
+		mappedProvider.provide(dependency, project, extension, postPopulationScheduler);
+	}
 
-				try (FileSystem fileSystem = FileSystems.newFileSystem(mappingsJar.toPath(), null)) {
-					Path fileToExtract = fileSystem.getPath("mappings/mappings.tiny");
-					Files.copy(fileToExtract, MAPPINGS_TINY_BASE.toPath());
-				}
-			}
+	private void storeMappings(Project project, MinecraftProvider minecraftProvider, File yarnJar) throws IOException {
+		if (isV2) {
+			// These are unmerged v2 mappings
+			Dependency intermediaryDep = project.getDependencies().add("compile", String.format("net.fabricmc:intermediary:%s:v2", minecraftProvider.minecraftVersion));
+			DependencyInfo intermediaryDepInfo = DependencyInfo.create(project, intermediaryDep, project.getConfigurations().getByName("compile"));
+			File intermediaryJar = intermediaryDepInfo.resolveFile().orElseThrow(() -> new RuntimeException("Could not find intermediaries: " + intermediaryDepInfo));
 
-			if (MAPPINGS_TINY.exists()) {
-				MAPPINGS_TINY.delete();
+			mergeAndSaveMappings(project, intermediaryJar.toPath(), yarnJar.toPath());
+		} else {
+			// These are merged v1 mappings
+			saveMergedV1Mappings(project, yarnJar.toPath());
+
+			if (tinyMappings.exists()) {
+				tinyMappings.delete();
 			}
 
 			project.getLogger().lifecycle(":populating field names");
-			new CommandProposeFieldNames().run(new String[]{minecraftProvider.MINECRAFT_MERGED_JAR.getAbsolutePath(), MAPPINGS_TINY_BASE.getAbsolutePath(), MAPPINGS_TINY.getAbsolutePath()});
+			suggestFieldNames(minecraftProvider, tinyMappingsWithoutEnums, tinyMappings.toPath());
 		}
+	}
 
-		mappedProvider = new MinecraftMappedProvider();
-		mappedProvider.initFiles(project, minecraftProvider, this);
-		mappedProvider.provide(dependency, project, extension, postPopulationScheduler);
+	private void saveMergedV1Mappings(Project project, Path mappingsJar) throws IOException {
+		project.getLogger().lifecycle(":extracting " + mappingsJar.getFileName());
```

Comment:
A lot of logging that should really be dropped down to info, only going to point this one out.